### ISSUE TITLE
depthwise_conv2d_native too slow

### ISSUE THREAD

**Issue by lubomir1** [tensorflow/tensorflow#19390]
***edit***: Simplified the example, added system info

According to this [thread](https://github.com/tensorflow/tensorflow/pull/17961) Tensorflow now uses the CuDNN accelerations of group convolutions for depthwise_conv2d_native. Thank you for working on this! However, I am having a hard time reproducing any gains from the accelerated version. Both the native and accelerated versions of depthwise_conv2d_native is about 3-4 times slower than doing a full (dense) convolution.

In the example below, a dense 3x3 convolution with 64-in and 64-out channels should do about 16 times more multiplications compared to a group convolution with the same dimensions and 16 groups (64x3x3x64 vs 16x4x3x3x4). The latter can be implemented with depthwise_conv2d_native with channel_multiplier of 4 followed by a sum. So I expect a fully amortized depthwise_conv2d_native to be 16 times faster than a dense convolution, and yet, it is about 4 times slower. It is also substantially slower than naive slice/convolve/concat implementation of group convolution

### System information
- **Have I written custom code**: no
- **OS Platform and Distribution**: Linux Ubuntu 16.04
- **TensorFlow installed from (source)**: from May 15, 2018 commit 1521eeb676383417b33ad55ad73b152bd5b046ca
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: 0.11.1
- **GCC/Compiler version (if compiling from source)**: 5.4
- **CUDA/cuDNN version**: 9.1, 7.1.3
- **GPU model and memory**: GeForce GTX 1080ti (tf compiled for compute capability 6.1) and Volta (V100, tf compiled for compute capability 7.0)
- **Exact command to reproduce**: see code below

Here are my results on GTX 1080ti:
```
           depthwise : 2.72s
     depthwise_cudnn : 2.59s
   manual_group_conv : 0.84s
          dense_conv : 0.72s
```
Here is the code I used to test performance:
```
import numpy as np, time, os
import tensorflow as tf
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

def conv3x3(bottom, filters):
    return tf.layers.conv2d(bottom, filters, kernel_size = 3, padding='same', use_bias = False, data_format = 'channels_first')

# this is not even group convolution, just the depthwise part without the sum
def depthwise(bottom, num_groups):
    input_chans = bottom.shape[1]
    group_size = input_chans / num_groups
    w0 = tf.get_variable(name='var', shape=[3, 3, input_chans, group_size])
    return tf.nn.depthwise_conv2d_native(bottom, w0, strides = [1,1,1,1], padding = 'SAME', data_format = 'NCHW')

def depthwise_cudnn(bottom, num_groups):
    with tf.get_default_graph()._kernel_label_map({"DepthwiseConv2dNative": "cudnn_grouped_convolution"}):
        return depthwise(bottom, num_groups)

def manual_group_conv(bottom, num_groups):
    input_chans = bottom.shape[1]
    group_size = input_chans / num_groups
    slices = [bottom[:,i:(i+group_size)] for i in range(0, input_chans, group_size)]
    convs = [conv3x3(sl, group_size) for sl in slices]
    return tf.concat(convs, axis = 1)

def dense_conv(bottom, num_groups):
    return conv3x3(bottom, bottom.shape[1])

input_shape = [16, 64, 128, 128]
groups = 16
dtype = tf.float32

for cnv_type in [depthwise, depthwise_cudnn, manual_group_conv, dense_conv]:
    tf.reset_default_graph()

    cnv = cnv_type(tf.constant(np.zeros(input_shape), dtype), groups)

    N = 100
    with tf.Session('') as sess:
        sess.run(tf.global_variables_initializer())

        sess.run(cnv) # initialization run

        start = time.time()
        for i in range(N):
            sess.run(cnv)

        print "%20s : %4.2fs" % (cnv_type.func_name, time.time() - start)

```

I also tested with NHWC with similar results. Let me know if I am doing something wrong or whether you can replicate my perf results. Thanks!

**Comment by lubomir1** [comment]
I updated accordingly

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@tfboyd could you take a look?

**Comment by drpngx** [comment]
@zheng-xq any idea?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@protoget do you know who would be able to handle that?

**Comment by protoget** [comment]
@chsigg Christian, I wonder if you worked on a custom kernel for depthwise conv before?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @drpngx, @tfboyd: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Robird** [comment]
same problem, need group conv
